### PR TITLE
Review of Service/Registration

### PR DESCRIPTION
**Pull request by pascalseeland**
Hi @smeyer-ilias  ,

I completed the review of the Services/Registration component.

Results:
- [x] The code is PSR-2 + X compliant. I did reformat some classes.
- [ ] No usage of $_GET / $_POST / $_REQUEST / $_SESSION outside GUI-classes
- [X] There is a Unit Test Suite with at least one unit test
- [X] The unit tests pass (they fullfil the formal criteria, but I would suggest adding some more usefull tests)
- [x] External libraries are compatible with PHP 8 (if relevant)
- [X] There are no serious PhpStorm Code Inspection issues left.
- [ ] The types are fully documented (PhpDoc) or explicit PHP types are used (type hints, return types, typed properties) (There was left I did not fix. The missing is did mark with TODO-PHP8-REVIEW .

Actions:
- [ ] Inline comment where prefixed with TODO-PHP8-REVIEW
- [ ] The EA-Plugin reports several warnings and one error. Please check them.

Kind regards,
Pascal

### CODE REVIEW

**Comment by mjansenDatabay** [comment]
@pascalseeland Hmmm, I already reviewed this  https://github.com/ILIAS-eLearning/ILIAS/pull/4225

File: Services/Registration/classes/class.ilAccountRegistrationGUI.php [hunk 1]
```diff
@@ -140,6 +147,8 @@ protected function __initForm() : void
         }
 
         // standard fields
+        //TODO-PHP8-REVIEW please check if there is a need for this static call. It looks like of odd to me, that
+        //we need a global static state variable in class that changes the behaviour of all instances.
```
Comment (smeyer-ilias) [inline Services/Registration/classes/class.ilAccountRegistrationGUI.php:151]: Yes, that's really odd. 
I will change this in Services/User.

File: Services/Registration/classes/class.ilAccountRegistrationGUI.php [hunk 2]
```diff
@@ -255,11 +264,11 @@ public function saveForm() : ilTemplate
                         $domains[] = $item;
                     }
                 }
-                if (sizeof($domains)) {
+                if (count($domains)) {
                     $mail_valid = false;
                     foreach ($domains as $domain) {
                         $domain = str_replace("*", "~~~", $domain);
-                        $domain = preg_quote($domain);
+                        $domain = preg_quote($domain); //TODO-PHP8-REVIEW please provide a delimiter, even if it is optional
```
Comment (smeyer-ilias) [inline Services/Registration/classes/class.ilAccountRegistrationGUI.php:271]: Ok, I will add the "/".

File: Services/Registration/classes/class.ilAccountRegistrationGUI.php [hunk 3]
```diff
@@ -460,7 +471,7 @@ protected function __createUser(int $a_role) : string
                             break;
 
                         case "relative":
-                            $rel = unserialize($code_data["alimitdt"]);
+                            $rel = unserialize($code_data["alimitdt"], ['allowed_classes' => false]);
```
Comment (smeyer-ilias) [inline Services/Registration/classes/class.ilAccountRegistrationGUI.php:474]: +1

File: Services/Registration/classes/class.ilRegistrationCode.php [hunk 4]
```diff
@@ -31,24 +41,24 @@ public static function create(
...
             $found = (bool) $ilDB->numRows($chk);
         }
 
-        if (is_array($local_roles)) {
-            $local_roles = implode(";", $local_roles);
-        }
-        if ($limit == "relative" && is_array($limit_date)) {
-            $limit_date = serialize($limit_date);
+        if ($limit === "relative" && is_array($limit_date)) {
+            $limit_date = serialize($limit_date); //TODO-PHP8-REVIEW please don't override variables with different types
```
Comment (smeyer-ilias) [inline Services/Registration/classes/class.ilRegistrationCode.php:53]: Ok

File: Services/Registration/classes/class.ilRegistrationSettingsGUI.php [hunk 5]
```diff
@@ -483,8 +474,8 @@ public function editRoleAccessLimitations(ilPropertyFormGUI $form = null) : void
             $this->lng->txt("registration_settings"),
             $this->ctrl->getLinkTarget($this, "view")
         );
-        $this->__initRoleAccessLimitations();
-        $form = $this->initRoleAccessForm();
+        $this->initRoleAccessLimitations();
+        $form = $this->initRoleAccessForm(); //TODO-PHP8-REVIEW this variable gets overwritten in any case.
```
Comment (smeyer-ilias) [inline Services/Registration/classes/class.ilRegistrationSettingsGUI.php:478]: Will fix this.

**Review by smeyer-ilias — APPROVED** [review]
Thanks @pascalseeland  for this detailed review and fixing issues. 

I will merge the PR and fix the remaing issues in the trunk.

**Comment by mjansenDatabay** [comment]
@smeyer-ilias Did you see my comment above related to PR #4225 ? This PR now has many conflicts *grml*

**Pull request merged**